### PR TITLE
WT-10688 Review chunk cache barrier usage

### PR DESCRIPTION
**Pull request by korteland**
Readers of `chunk->valid` need to ensure that reading other `chunk` fields happens strictly after checking validity, and writers need to ensure that it is the last field written to.

Unfortunately locking doesn't really solve the problem here - this reordering can still happen while holding the bucket lock, and there can be a large number of chunks so finer-grained locking just for `valid` field would be wasteful. Locking is also likely to have a measurable performance impact - significant work has gone into minimising the critical section, and since the lock is at the bucket level, reading or writing `valid` with the bucket lock will prevent anyone else looking at the entire bucket.

Use `ORDERED_READ`/`PUBLISH` as the longer-term solution of acquire/release isn't ready yet.

### CODE REVIEW

File: src/block_cache/block_chunkcache.c
Comment (Siddhartha8899) [inline src/block_cache/block_chunkcache.c:600]: Nit: As per the WT convention, shouldn't primitive variables be declared at the end?

**Comment by luke-pearson** [comment]
>  this reordering can still happen while holding the bucket lock,
Can you be explicit and give an example of what reordering you are referring to? Typically locking would prevent reordering.

File: src/block_cache/block_chunkcache.c
Comment (korteland) [reply src/block_cache/block_chunkcache.c]: Oops, I was a bit too vigorous in where I asked Emacs to sort lines - thanks. Fixed in fee8306

**Comment by korteland** [comment]
> Can you be explicit and give an example of what reordering you are referring to? Typically locking would prevent reordering.

Sure - in `__wt_chunkcache_get`, we had this:
```
if (!chunk->valid) {
    // snip - retry logic
}
/* Found the needed chunk. */
chunk_cached = true;
WT_ASSERT(...);

/* We can't read beyond the chunk's boundary. */
readable_in_chunk =
  (size_t)chunk->chunk_offset + chunk->chunk_size - (size_t)offset;
```

If the CPU or compiler were to read `chunk->chunk_offset` or `chunk->chunk_size` before valid, it might find that they aren't set yet, and at that point it would clearly be a logic error. And the same applies to the write thread, but in the opposite order.

**Comment by luke-pearson** [comment]
> If the CPU or compiler were to reorder the read of `chunk->chunk_offset` or `chunk->chunk_size` before `chunk->valid`, it might find that they aren't set yet, and at that point it would clearly be a logic error. The write thread has an analogous problem around ensuring `chunk->valid` is written last.

I think you've snipped out some important logic, is that the retry that has the `spin_unlock` call in it? Your usage of barriers looks correct but I'm still not 100% convinced we need them.

**Comment by korteland** [comment]
> I think you've snipped out some important logic, is that the retry that has the spin_unlock call in it?

It is, but I don't agree it's important. The chunk being valid is expected to be the common case - the only time we'd find a chunk but have `!valid` is when there's IO in progress during the initial insertion of the chunk into the cache. So in practice it's relatively rare we hit the unlock.

I can see the argument that the compiler wouldn't move anything past the unlock, but I don't see how the CPU would know not to reorder reads there.

> Your usage of barriers looks correct but I'm still not 100% convinced we need them.

What would be the bar to meet here?

**Comment by luke-pearson** [comment]
> I can see the argument that the compiler wouldn't move anything past the unlock, but I don't see how the CPU would know not to reorder reads there.

The CPU can't move things across a lock either however that lock is sitting in an if statement, I don't see a mechanism to block the CPU. 

> What would be the bar to meet here?

This discussion is sufficient, I am now more convinced that you will need the barriers. I just wanted to understand the lock relevancy. @ajmorton will also need to sign-off. So please wait for his review as well.

**Comment by ajmorton** [comment]
I'm happy with the barriers. There's a chance they're not needed here but I don't want to dig into the subtleties of mixing `if`s, `goto`s, and `return`s to figure that out.
Adding these barriers is also an established pattern in WiredTiger which we're currently calling `WT_FLAG_PUBLISH` in our docs. Ensuring that the "ready to read" flag is written last and read first is good practice.

@luke-pearson are we still planning to review this code in PM-3221? We'll need to convert these ORDERED_READS into READ_ONCE/READ_ORDEREDs at some point.

File: src/block_cache/block_chunkcache.c
Comment (ajmorton) [inline src/block_cache/block_chunkcache.c:577]: Can we add a comment here that the `PUBLISH` and `ORDERED_READ`s are required since populating the chunk isn't protected by the bucket_lock?

File: src/block_cache/block_chunkcache.c
Comment (korteland) [reply src/block_cache/block_chunkcache.c:290]: Yep - @Siddhartha8899 pointed out that we don't want this underflowing.

**Comment by luke-pearson** [comment]
> @luke-pearson are we still planning to review this code in PM-3221?

We should track the new barriers but no I don't think chunk cache will fall under our purview

File: src/block_cache/block_chunkcache.c
Comment (korteland) [reply src/block_cache/block_chunkcache.c:582]: Good call - done in 8334956.

File: src/block_cache/block_chunkcache.c
Comment (ajmorton) [reply src/block_cache/block_chunkcache.c:295]: I think we need an extra case for `chunk->access_count == 1` then in that case.
The previous behavior was to decrement and return true, now we're decrementing and returning false which delays eviction until the subsequent `should_evict` call.

Comment (korteland) [reply src/block_cache/block_chunkcache.c:295]: If we wanted identical behaviour, yeah. It's not important to match the old behaviour 100%, eviction isn't on the critical path and it's still the case that less-referenced chunks will get evicted first.

(It's a clock algorithm - the eviction server sits there spinning over the table decrementing, and chunk reads bump this count. So this change is functionally the same as if each chunk had been accessed one extra time.)

**Review by ruby-oujo — APPROVED** [review]

**Review by ajmorton — APPROVED** [review]
LGTM 👍

**Review by luke-pearson — APPROVED** [review]
LGTM. Reviewed the barrier usage.

**Review by Siddhartha8899 — APPROVED** [review]

**Comment by korteland** [comment]
evergreen merge

**Pull request merged**
